src/js-utils.c

```diff
@@ -0,0 +1,117 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/*
+ * js-utils.c
+ *
+ * Copyright (C) 2018 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Author: Juan Pablo Ugarte <ugarte@endlessm.com>
+ *
+ */
+
+#include "js-utils.h"
+
+static void
+_js_run_finish_handler (GObject *object, GAsyncResult *result, gpointer data)
+{
+  WebKitJavascriptResult *js_result;
+  gchar *function = data;
+  GError *error = NULL;
+
+  js_result = webkit_web_view_run_javascript_finish (WEBKIT_WEB_VIEW (object),
+                                                     result, &error);
+  if (!js_result)
+    {
+      g_warning ("JS Error in %s(): %s", function, error->message);
+      g_error_free (error);
+    }
+}
+
+void
+_js_run_printf (WebKitWebView *webview,
+                const gchar   *function,
+                const gchar   *format,
+                ...)
+{
+  gchar *script;
+  va_list args;
+
+  va_start (args, format);
+  script = g_strdup_vprintf (format, args);
+  va_end (args);
+
+  webkit_web_view_run_javascript (WEBKIT_WEB_VIEW (webview),
+                                  script, NULL,
+                                  _js_run_finish_handler,
+                                  (gpointer)function);
```

Comment:
You're not guaranteed that the memory pointed to by `function` will still be valid by the time the async function has returned, so you should probably copy the string before passing it (and then freeing it in the callback)

Reply:
Ohh right, __func__ is not a literal!